test/e2e/helpers_test.go

```diff
@@ -0,0 +1,284 @@
+	}
+	var memberList etcdserverpb.MemberListResponse
+	if err := json.Unmarshal([]byte(stdout), &memberList); err != nil {
+		t.Fatalf("Failed to parse etcd member list JSON: %v", err)
+	}
+	return &memberList
+}
+
+// verifyPodUsesPVC checks that a pod is using PVC for persistent storage
+func verifyPodUsesPVC(t *testing.T, c *envconf.Config, podName string, expectedPVCPrefix string) {
+	t.Helper()
+	var pod corev1.Pod
+	if err := c.Client().Resources().Get(t.Context(), podName, namespace, &pod); err != nil {
+		t.Fatalf("Failed to get pod %s: %v", podName, err)
+	}
+
+	// Check for PVC volumes
+	hasPVC := false
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim != nil {
+			if strings.HasPrefix(volume.PersistentVolumeClaim.ClaimName, expectedPVCPrefix) {
+				hasPVC = true
+				break
+			}
+		}
+	}
+
+	if !hasPVC {
+		t.Errorf("Pod %s does not use expected PVC with prefix %s", podName, expectedPVCPrefix)
+	}
+}
+
+// getLocalEndpointHashKV executes `etcdctl endpoint hashkv -w json` inside the given pod
+// and returns the parsed etcdserverpb.HashKVResponse from the local member.
+//
+// Supported JSON shapes (based on etcdctl versions):
+//  1. Single object: {"header":{...},"hash":...,"compact_revision":...,"hash_revision":...}
+//  2. Array with nested object: [{"Endpoint":"...","HashKV":{...}}]
+//  3. Array with flat fields:
+//     [{"Endpoint":"...", "hash":..., "compact_revision":...,
+//     "hash_revision":..., "header":{...}}]
+func getLocalEndpointHashKV(t *testing.T, c *envconf.Config, podName string) *etcdserverpb.HashKVResponse {
+	t.Helper()
+	stdout, stderr, err := execInPod(t, c, podName, namespace, []string{"etcdctl", "endpoint", "hashkv", "-w", "json"})
+	if err != nil {
+		t.Fatalf("Failed to get endpoint hashkv from %s: %v, stderr: %s", podName, err, stderr)
+	}
+
+	// Case 1: direct single-object response
+	var direct etcdserverpb.HashKVResponse
+	if err := json.Unmarshal([]byte(stdout), &direct); err == nil &&
+		(direct.Hash != 0 || direct.HashRevision != 0 ||
+			direct.CompactRevision != 0 || direct.Header != nil) {
+		return &direct
+	}
+
+	// Case 2/3: array responses
+	var elems []json.RawMessage
+	if err := json.Unmarshal([]byte(stdout), &elems); err == nil && len(elems) > 0 {
+		// 2) Nested HashKV under capitalized key
+		var nested struct {
+			Endpoint string                      `json:"Endpoint"`
+			HashKV   etcdserverpb.HashKVResponse `json:"HashKV"`
+		}
+		if err := json.Unmarshal(elems[0], &nested); err == nil &&
+			(nested.HashKV.Hash != 0 || nested.HashKV.HashRevision != 0 ||
+				nested.HashKV.CompactRevision != 0 || nested.HashKV.Header != nil) {
+			return &nested.HashKV
+		}
+		// 3) Flat fields at element level (unknown keys ignored)
+		var flat etcdserverpb.HashKVResponse
+		if err := json.Unmarshal(elems[0], &flat); err == nil &&
+			(flat.Hash != 0 || flat.HashRevision != 0 ||
+				flat.CompactRevision != 0 || flat.Header != nil) {
+			return &flat
+		}
+	}
```

Comment:
Why it's so complicated to parse the response? If you specify `--cluster`, then you will get all members' responses, otherwise it's just one member's response.

Reply:
Thanks for the reminder. I've updated the usage with `--cluster` and simplified parsing.